### PR TITLE
build(build-tools): Upgrade eslint config

### PR DESCRIPTION
**Pull request by tylerbutler**
This PR changes the following:

1. Updates the eslint config used in build-tools to the latest version.
2. Use the recommended config in the build-cli project
3. Enable noImplicitAny and noUnusedLocals in the build-cli tsconfig.

### CODE REVIEW

File: build-tools/packages/build-cli/.eslintrc.cjs [hunk 1]
```diff
@@ -9,12 +9,12 @@ module.exports = {
...
 		// eslint-disable-next-line node/no-extraneous-require
-		require.resolve("@fluidframework/eslint-config-fluid/minimal"),
+		require.resolve("@fluidframework/eslint-config-fluid/recommended"),
 		"prettier",
 	],
 	rules: {
-		"@typescript-eslint/no-unused-vars": "warn",
-		"unused-imports/no-unused-imports": "warn",
+		// "@typescript-eslint/no-unused-vars": "warn",
+		// "unused-imports/no-unused-imports": "warn",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/.eslintrc.cjs:18]:
```suggestion

```

File: build-tools/packages/build-tools/.eslintrc.js [hunk 2]
```diff
@@ -11,10 +11,17 @@ module.exports = {
...
-		"@typescript-eslint/no-var-requires": "off",
+		// TODO: Enable these ASAP
+		"@typescript-eslint/no-explicit-any": "off",
+		"@typescript-eslint/no-unsafe-argument": "off",
+		"@typescript-eslint/no-unsafe-assignment": "off",
+		"@typescript-eslint/no-unsafe-call": "off",
+		"@typescript-eslint/no-unsafe-member-access": "off",
+		"@typescript-eslint/no-unused-vars": "warn",
+		//"@typescript-eslint/no-inferrable-types": "off",
+		//"@typescript-eslint/no-var-requires": "off",
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/.eslintrc.js:22]:
```suggestion
```

File: build-tools/packages/bundle-size-tools/package.json [hunk 3]
```diff
@@ -39,18 +39,18 @@
...
 		"copyfiles": "^2.4.1",
-		"eslint": "~8.6.0",
+		"eslint": "~8.51.0",
 		"eslint-plugin-eslint-comments": "~3.2.0",
-		"eslint-plugin-import": "~2.25.4",
-		"eslint-plugin-unicorn": "~40.0.0",
-		"eslint-plugin-unused-imports": "~2.0.0",
+		"eslint-plugin-import": "~2.28.1",
+		"eslint-plugin-unicorn": "~48.0.1",
+		"eslint-plugin-unused-imports": "~3.0.0",
```
Comment (tylerbutler) [inline build-tools/packages/bundle-size-tools/package.json:53]:
```suggestion
```

File: build-tools/packages/bundle-size-tools/tsconfig.json [hunk 4]
```diff
@@ -5,11 +5,12 @@
...
-		"strictBindCallApply": true,
-		"strictFunctionTypes": true,
-		"strictNullChecks": true,
-		"strictPropertyInitialization": true,
+		"noImplicitAny": true,
+		// "noUnusedLocals": false,
+		// "strictBindCallApply": true,
+		// "strictFunctionTypes": true,
+		// "strictNullChecks": true,
+		// "strictPropertyInitialization": true,
```
Comment (tylerbutler) [inline build-tools/packages/bundle-size-tools/tsconfig.json:13]:
```suggestion
```

File: build-tools/packages/build-cli/src/commands/bump.ts [hunk 5]
```diff
@@ -48,6 +47,7 @@ export default class BumpCommand extends BaseCommand<typeof BumpCommand> {
 			description:
 				"Bump the release group or package to the next version according to this bump type.",
 			exclusive: ["exact"],
+			multiple: false,
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/bump.ts:50]:
```suggestion
```

File: build-tools/packages/build-cli/src/commands/bump.ts [hunk 6]
```diff
@@ -58,6 +58,7 @@ export default class BumpCommand extends BaseCommand<typeof BumpCommand> {
 			description: "Override the version scheme used by the release group or package.",
 			required: false,
 			exclusive: ["exact"],
+			multiple: false,
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/bump.ts:61]:
```suggestion
```

File: build-tools/packages/build-cli/src/commands/bump/deps.ts [hunk 7]
```diff
@@ -52,6 +52,7 @@ export default class DepsCommand extends BaseCommand<typeof DepsCommand> {
 			char: "t",
 			default: "minor",
 			description: "Bump the current version of the dependency according to this bump type.",
+			multiple: false,
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/bump/deps.ts:55]:
```suggestion
```

File: build-tools/packages/build-cli/src/commands/release.ts [hunk 8]
```diff
@@ -55,18 +55,19 @@ export default class ReleaseCommand extends StateMachineCommand<typeof ReleaseCo
 			required: false,
 		}),
 		bumpType: bumpTypeFlag({
+			multiple: false,
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/release.ts:58]:
```suggestion
```

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 9]
```diff
@@ -559,11 +562,9 @@ export async function setVersion(
 	if (prettierConfig !== null) {
 		prettierConfig.filepath = lernaPath;
 	}
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
 	lernaJson.version = translatedVersion.version;
-	const output = await prettier(
-		JSON.stringify(lernaJson),
-		prettierConfig === null ? undefined : prettierConfig,
-	);
+	const output = await prettier(JSON.stringify(lernaJson), prettierConfig ?? undefined);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/lib/package.ts:567]:
```suggestion
	const output = await prettier(
		JSON.stringify(lernaJson),
		// eslint-disable-next-line @typescript-eslint/prefer-null-coalescing
		prettierConfig === null ? undefined : prettierConfig,
	);
```

File: build-tools/packages/bundle-size-tools/package.json [hunk 10]
```diff
@@ -39,18 +39,15 @@
...
+		"@fluidframework/eslint-config-fluid": "^3.0.0",
 		"@microsoft/api-extractor": "^7.36.1",
 		"@types/msgpack-lite": "^0.1.8",
 		"@types/node": "^14.18.53",
 		"@types/pako": "^2.0.0",
 		"concurrently": "^8.2.1",
 		"copyfiles": "^2.4.1",
-		"eslint": "~8.6.0",
+		"eslint": "~8.51.0",
 		"eslint-plugin-eslint-comments": "~3.2.0",
```
Comment (tylerbutler) [inline build-tools/packages/bundle-size-tools/package.json:50]:
```suggestion
```

File: build-tools/packages/bundle-size-tools/tsconfig.json [hunk 11]
```diff
@@ -5,11 +5,7 @@
 		"rootDir": "./src",
 		"composite": true,
 		"noImplicitThis": true,
-		"noUnusedLocals": false,
-		"strictBindCallApply": true,
-		"strictFunctionTypes": true,
-		"strictNullChecks": true,
-		"strictPropertyInitialization": true,
```
Comment (tylerbutler) [inline build-tools/packages/bundle-size-tools/tsconfig.json:12]: All default values from the base config.

File: build-tools/packages/build-cli/.eslintrc.cjs [hunk 12]
```diff
@@ -9,12 +9,12 @@ module.exports = {
...
 		"oclif-typescript",
 		// eslint-disable-next-line node/no-extraneous-require
-		require.resolve("@fluidframework/eslint-config-fluid/minimal"),
+		require.resolve("@fluidframework/eslint-config-fluid/recommended"),
 		"prettier",
 	],
 	rules: {
-		"@typescript-eslint/no-unused-vars": "warn",
-		"unused-imports/no-unused-imports": "warn",
+		// This rule is often triggered when using custom Flag classes, so disabling.
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/.eslintrc.cjs:16]:
```suggestion
		// This rule is often triggered when using custom Flags, so disabling.
```

File: build-tools/packages/build-cli/.eslintrc.cjs [hunk 13]
```diff
@@ -50,6 +50,9 @@ module.exports = {
 		// Too strict for our needs
 		"unicorn/filename-case": "off",
 
+		// Destructuring is often awkward to use with commands
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/.eslintrc.cjs:53]:
```suggestion
		// In commands, destructuring is useful in some places but makes others less legible, so consistency isn't preferred.
```

File: build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/tscTask.ts [hunk 14]
```diff
@@ -13,6 +13,7 @@ import { getInstalledPackageVersion } from "../../../common/taskUtils";
 import { getTscUtils, TscUtil } from "../../../common/tscUtils";
 import { LeafTask, LeafWithDoneFileTask } from "./leafTask";
 
+// eslint-disable-next-line @typescript-eslint/no-var-requires
 const isEqual = require("lodash.isequal");
```
Comment (curtisman) [inline build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/tscTask.ts:17]: I wonder why we need need var require here?

**Review by curtisman — APPROVED** [review]

**Pull request merged**